sequencers/single/single.go

```diff
@@ -0,0 +1,402 @@
+	}
+}
+
+// submitBatchToDA submits a batch of transactions to the Data Availability (DA) layer.
+// It implements a retry mechanism with exponential backoff and gas price adjustments
+// to handle various failure scenarios.
+//
+// The function attempts to submit the batch multiple times (up to maxSubmitAttempts)
+// with different strategies based on the response from the DA layer:
+// - On success: Reduces gas price gradually (but not below initial price)
+// - On mempool issues: Increases gas price and uses a longer backoff
+// - On size issues: Reduces the blob size and uses exponential backoff
+// - On other errors: Uses exponential backoff
+//
+// It returns an error if not all batches could be submitted after all attempts.
+func (c *Sequencer) submitBatchToDA(ctx context.Context, batch coresequencer.Batch) error {
+	// Initialize a slice with the batch to submit
+	batchesToSubmit := []*coresequencer.Batch{&batch}
+	submittedAllBlocks := false
+	var backoff time.Duration
+	numSubmittedBatches := 0
+	attempt := 0
+
+	// Store initial values to be able to reset or compare later
+	maxBlobSize := c.maxBlobSize.Load()
+	initialMaxBlobSize := maxBlobSize
+	initialGasPrice, err := c.dalc.GasPrice(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get initial gas price: %w", err)
+	}
+	gasPrice := initialGasPrice
+
+daSubmitRetryLoop:
+	for !submittedAllBlocks && attempt < maxSubmitAttempts {
+		// Wait for backoff duration or exit if context is done
+		select {
+		case <-ctx.Done():
+			break daSubmitRetryLoop
+		case <-time.After(backoff):
+		}
+
+		data := make([][]byte, len(batchesToSubmit))
+		for i, batch := range batchesToSubmit {
+			// transform batchesToSubmit into [][]byte
+			pbBatch := pb.Batch{Txs: batch.Transactions}
+			bz, err := pbBatch.Marshal()
+			if err != nil {
+				return fmt.Errorf("failed to marshal batch: %w", err)
+			}
+			data[i] = bz
+		}
+		// Attempt to submit the batch to the DA layer
+		res := c.dalc.SubmitBatch(ctx, data, maxBlobSize, gasPrice)
+
+		gasMultiplier, err := c.dalc.GasMultiplier(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get gas multiplier: %w", err)
+		}
+
+		switch res.Code {
+		case coreda.StatusSuccess:
+			//TODO: upon success, mark the batch as processed in the WAL & update the last submitted batch hash
+			// Count total transactions for logging
+			txCount := 0
+			for _, batch := range batchesToSubmit {
+				txCount += len(batch.Transactions)
+			}
+			c.logger.Info("successfully submitted batches to DA layer", "gasPrice", gasPrice, "height", res.Height, "batchCount", res.SubmittedCount, "txCount", txCount)
+
+			// Check if all batches were submitted
+			if res.SubmittedCount == uint64(len(batchesToSubmit)) {
+				submittedAllBlocks = true
+			}
+
+			// Split the batches into submitted and not submitted
+			submittedBatches, notSubmittedBatches := batchesToSubmit[:res.SubmittedCount], batchesToSubmit[res.SubmittedCount:]
```

Comment:
this logic needs to be updated. the `SubmitBatch` is operating within a single batch and it will return the number of txs that it was able to submit as blobs. The remaining txs from the single batch needs to be extracted out and resubmitted. This logic is assuming multiple batches (see batchesToSubmit declaration in line 209).